app/controllers/concerns/eventable.rb

```diff
@@ -8,31 +8,41 @@ def supplier_id
 
   def process_event(eventables, event_name, event_params)
     [eventables].flatten.each do |eventable|
-      event = create_event(eventable, event_name, event_params)
-      create_generic_event(event)
+      create_generic_event(eventable, event_name, event_params)
       run_event_logs(eventable)
     end
   end
 
-  def create_event(eventable, event_name, event_params)
-    eventable.events.create!(
-      event_name: event_name,
-      client_timestamp: Time.zone.parse(event_params.dig(:attributes, :timestamp)),
-      details: {
-        event_params: event_params,
-        supplier_id: supplier_id,
-      },
-    )
-  end
+  def create_generic_event(eventable, event_name, event_params)
+    eventable_type = eventable.class.to_s
+    type = "GenericEvent::#{eventable_type}#{event_name.camelize}".constantize
+
+    attributes = {}
+    assign_common_attributes!(attributes, eventable, event_params)
+    assign_specific_attributes!(attributes, event_params, type)
 
-  def create_generic_event(event)
-    generic_event = GenericEvent.from_event(event)
-    generic_event.save!
-    event.update!(generic_event_id: generic_event.id)
-    generic_event
+    type.create!(attributes)
   end
 
   def run_event_logs(eventable)
     GenericEvents::Runner.new(eventable).call
   end
+
+private
+
+  def assign_common_attributes!(attributes, eventable, event_params)
+    timestamp = Time.zone.parse(event_params.dig(:attributes, :timestamp))
+
+    attributes[:eventable]   = eventable
+    attributes[:occurred_at] = timestamp
```

Comment:
how about using attributes `occured_at` and `recorded_at` and fall back to `timestamp` if they are not provided?

Reply:
This module is only used in an api where timestamp is provided so I'm not sure the flexibility adds value? I think our intention is probably to leisurely migrate all calls to be via `POST /api/events` rather than these existing endpoints and ultimately deprecate and then remove this code.

Reply:
ah of course, sorry for not realising that 👍